firedrake/assemble.py

```diff
@@ -425,10 +426,31 @@ def allocate_matrix(expr, bcs=None, *, mat_type=None, sub_mat_type=None,
                                          for get_map, regions in domains.items()
                                          if regions))
     try:
+        # Let tensor=None and bcs=None for the purpose of allocation.
+        parloop_builders = TwoFormAssembler(expr, None, bcs=None, form_compiler_parameters=form_compiler_parameters).parloop_builders_for_sparsity_construction
+        maps_and_regions = defaultdict(lambda: defaultdict(set))
+        if any(parloop_builder._indices == (None, None) for parloop_builder in parloop_builders):
+            # Handle special cases: slate or split=False
+            assert all(parloop_builder._indices == (None, None) for parloop_builder in parloop_builders)
+            for parloop_builder in parloop_builders:
+                _integral_type = parloop_builder._integral_type
+                get_map, region = mapping[_integral_type]
+                for i, _rmap in enumerate(get_map(test)):
+                    for j, _cmap in enumerate(get_map(trial)):
+                        maps_and_regions[(i, j)][(_rmap, _cmap)].update((region, ))
+        else:
+            for parloop_builder in parloop_builders:
+                i, j = parloop_builder._indices
+                _integral_type = parloop_builder._integral_type
+                # Make Sparsity independent of _iterset for better reusability.
+                get_map, region = mapping[_integral_type]
+                _rmap = get_map(test).split[i] if get_map(test) is not None else None
+                _cmap = get_map(trial).split[j] if get_map(trial) is not None else None
+                maps_and_regions[(i, j)][(_rmap, _cmap)].update((region, ))
+        maps_and_regions = {key: [k + (tuple(v), ) for k, v in val.items()] for key, val in maps_and_regions.items()}
```

Comment:
This line is very obscure. Could it be made clearer?

Reply:
Fixed.